src/operator/nn/activation-inl.h

```diff
@@ -0,0 +1,172 @@
+    Assign(out, req, F<ForwardOp>(data));
+  }
+
+  virtual void Backward(const OpContext &ctx, const TBlob &out_grad,
+                        const TBlob &out_data, const OpReqType &req,
+                        const TBlob &in_grad) {
+    using namespace mshadow;
+    using namespace mshadow::expr;
+    Stream<xpu> *s = ctx.get_stream<xpu>();
+    Tensor<xpu, 2, DType> m_out_grad = out_grad.FlatTo2D<xpu, DType>(s);
+    Tensor<xpu, 2, DType> m_out_data = out_data.FlatTo2D<xpu, DType>(s);
+    Tensor<xpu, 2, DType> m_in_grad = in_grad.FlatTo2D<xpu, DType>(s);
+    Assign(m_in_grad, req, F<BackwardOp>(m_out_data) * m_out_grad);
+  }
+};  // class ActivationOp
+
+template<typename xpu, typename ForwardOp, typename BackwardOp, typename DType>
+ActivationOp<xpu, ForwardOp, BackwardOp, DType> &get_activation_op()
+{
+  static thread_local ActivationOp<xpu, ForwardOp, BackwardOp, DType> op;
+  return op;
+}
+
+template<typename xpu>
+void ActivationCompute(const nnvm::NodeAttrs& attrs,
+    const OpContext& ctx,
+    const std::vector<TBlob>& inputs,
+    const std::vector<OpReqType>& req,
+    const std::vector<TBlob>& outputs) {
+  CHECK_EQ(inputs.size(), 1U);
+  CHECK_EQ(outputs.size(), 1U);
+  const ActivationParam& param = nnvm::get<ActivationParam>(attrs.parsed);
+  MSHADOW_REAL_TYPE_SWITCH(inputs[0].type_flag_, DType, {
+    switch (param.act_type) {
+      case activation::kReLU:
+        get_activation_op<xpu, mshadow_op::relu, mshadow_op::relu_grad, DType>().Forward(
+            ctx, inputs[0], req[0], outputs[0]);
+        break;
+      case activation::kSigmoid:
+        get_activation_op<xpu, mshadow_op::sigmoid, mshadow_op::sigmoid_grad, DType>().Forward(
+            ctx, inputs[0], req[0], outputs[0]);
+        break;
+      case activation::kTanh:
+        get_activation_op<xpu, mshadow_op::tanh, mshadow_op::tanh_grad, DType>().Forward(
+            ctx, inputs[0], req[0], outputs[0]);
+        break;
+      case activation::kSoftReLU:
+        get_activation_op<xpu, mshadow_op::softrelu, mshadow_op::softrelu_grad, DType>().Forward(
+            ctx, inputs[0], req[0], outputs[0]);
+        break;
+      default:
+        LOG(FATAL) << "unknown activation type";
+    }
+  });
+}
+
+template<typename xpu>
+void ActivationGradCompute(const nnvm::NodeAttrs& attrs,
+    const OpContext& ctx,
+    const std::vector<TBlob>& inputs,
+    const std::vector<OpReqType>& req,
+    const std::vector<TBlob>& outputs) {
+#if MXNET_USE_CUDNN == 1
+  CHECK_EQ(inputs.size(), 3U);
+#else
+  CHECK_EQ(inputs.size(), 2U);
+#endif
+  CHECK_EQ(outputs.size(), 1U);
+  CHECK_EQ(req.size(), 1U);
+  const ActivationParam& param = nnvm::get<ActivationParam>(attrs.parsed);
+  MSHADOW_REAL_TYPE_SWITCH(inputs[0].type_flag_, DType, {
+    switch (param.act_type) {
+      case activation::kReLU:
+        get_activation_op<xpu, mshadow_op::relu, mshadow_op::relu_grad, DType>().Backward(
+            ctx, inputs[0], inputs[1], req[0], outputs[0]);
```

Comment:
Why not just make this Forward()/Backward() function a static member function and call them directly rather than calling some thread_local static object?  It seems to adding a several levels of complication that don't need to exist.

Reply:
Agreed. I was trying to reduce the number of lines of code moved as much as possible. Although the code looks a little weird, it doesn't have performance cost. We still need to maintain a bunch of states and initialize them whenever forward/backward is called.

Reply:
What states? Just the params?